spatial-tracking-explainer.md

```diff
@@ -0,0 +1,382 @@
+
+### Frame of Reference Examples
+
+| Type                           | Subtype             | Example                                       |
+| -------------------------------| ------------------- | --------------------------------------------- |
+| `XRStationaryFrameOfReference` | `position-disabled` | 360 photo/video viewer                        |
+| `XRStationaryFrameOfReference` | `eye-level`         | Racing game or space simulator                |
+| `XRStationaryFrameOfReference` | `floor-level`       | Action game where you duck and dodge in place |
+| `XRBoundedFrameOfReference`    |                     | Room redecorator                              |
+| `XRUnboundedFrameOfReference`  |                     | Walking tour                                  |
+
+### XRFrameOfReference Availability
+
+**Guaranteed** The UA will always be able to provide this frame of reference 
+
+**Hardware-dependent** The UA will only be able to supply this frame of reference if running on XR hardware that supports it
+
+**Rejected** The UA will never provide this frame of reference
+
+&ast; Or polyfillable on inside-out tracking systems with anchor support.  See the section on "Floor Alignment Compatibility".
+
+| Type                           | Subtype             | Inline             | Immersive |
+| -------------------------------| ------------------- | ------------------ | ---------- |
+| `XRStationaryFrameOfReference` | `position-disabled` | Hardware-dependent | Guaranteed |
+| `XRStationaryFrameOfReference` | `eye-level`         | Hardware-dependent | Guaranteed |
+| `XRStationaryFrameOfReference` | `floor-level`       | Hardware-dependent | Guaranteed* |
+| `XRBoundedFrameOfReference`    |                     | Rejected           | Hardware-dependent* |
+| `XRUnboundedFrameOfReference`  |                     | Rejected           | Hardware-dependent |
+
+## Appendix B: Proposed partial IDL
+This is a partial IDL and is considered additive to the core IDL found in the main [explainer](explainer.md).
+
+```webidl
+//
+// Session
+//
+
+partial dictionary XRSessionCreationOptions {
+  XRFrameOfReferenceType requiredFrameOfReferenceType;
+};
+
+partial interface XRSession {
+  Promise<XRFrameOfReference> requestFrameOfReference(XRFrameOfReferenceOptions options);
+};
+
+//
+// Frames and Poses
+//
+
+partial interface XRFrame {
+  // Also listed in the main explainer.md
+  XRDevicePose? getDevicePose(optional XRFrameOfReference frameOfReference);
+  XRInputPose? getInputPose(XRInputSource inputSource, optional XRFrameOfReference frameOfReference);
+};
+
+[SecureContext, Exposed=Window]
+interface XRInputPose {
+  readonly attribute boolean emulatedPosition;
+  readonly attribute XRRay targetRay;
+  readonly attribute Float32Array? gripMatrix;
+};
+
+//
+// Coordinate System
+//
+
+[SecureContext, Exposed=Window] interface XRCoordinateSystem {
+  Float32Array? getTransformTo(XRCoordinateSystem other);
```

Comment:
Can we get a code example of how this is used?

Reply:
Yeah, this is a tricky one!  I *almost* left out the `XRCoordinateSystem` type entirely in this PR because it is pretty meaningless until we have anchors and similar types.  (which is why that's next on my list!)  The main reason I left it was that I didn't want to give the impression that it was being removed permanently when most of the rest of the PR was aiming for parity (but with better clarity).

That said, since a few folks have missed it when reading...it's worth noting two things:
1. `XRFrameOfReference` no longer inherits from `XRCoordinateSystem`.  It has a member instead.  This pattern is one that seems like it will adapt better to the inclusion of anchors, meshes, and the like.  And 
1. The `getDevicePose()` and `getInputPose()` no longer take `XRCoordinateSystem` objects, but instead take `XRFrameOfReference` objects.  This more strongly correlates the mental model that these poses are only really relevant in the context of the `XRFrameOfReference` being used for rendering.

All that said... to answer your questions:
1. It should be possible so long as the two `XRFrameOfReference` objects are able to be related to one another at the moment the transition is desired to happen.  For example, if I put a bag over my head, walked down the hall, and created a new instance... well those probably won't be relatable until I walk back down the hallway bag-less.
1. I'm not entirely sure what you mean by "Would it make more sense to have the content from the old type be attached to a XRCoordinateSystem that came from the difference between the two types?" Are you suggesting an `XRCoordinateSystem` can be "new'd" up?
1. Yes ... but it will involve anchors ^_^;;;  And we're back to the problem of needing to put a cut line *somehwere* to get a PR out the door.  I'll be thrilled to get the anchors-related examples up because it'll be a much clearer explanation.

Reply:
For question 3, the type that would represent a plane, mesh, free-space anchor or other tracked element would provide an (or be an) `XRCoordinateSystem`, allowing you to relate its pose to any frame of reference or any other such element.

We could alternatively give all such world entities a `.getPose` method that returns its current transform in a developer's preferred frame of reference, without having an `XRCoordinateSystem` for each such object.  However, once you give that method an `XRCoordinateSystem` parameter so the developer can specify the base frame they want, you've got the same signature as `.getTransformTo` again, just one type up.

Reply:
@lincolnfrog Not sure if my question above got lost in the deluge of PR comments.  What were you meaning by 

>Would it make more sense to have the content from the old type be attached to a XRCoordinateSystem that came from the difference between the two types?

Are you suggesting an XRCoordinateSystem can be "new'd" up?

Reply:
I guess that part of my question does assume that XRCoordinateSystems could be created arbitrarily - and that you could get the transform between the new FOR and the old FOR and then create a XRCS that represents that transform. You could then attach all your content from the old FOR to that XRCS so that it renders correctly when you get the device pose using the new FOR.

Did that paragraph make ANY sense? I may be fundamentally misunderstanding something here, so hopefully, at the very least, my confusion can help.

The ultimate point I am trying to understand here is what the approach looks like to transition virtual content from an unbounded FOR situation to a bounded FOR situation and still have it render in "approximately" the right location. I believe it is important to either say "you can't do that" or to make sure it's something we support without requiring multiple render passes where the app developer has to get two FORs and use each one in a camera to render the appropriate scene.

Reply:
Chatted with Max and have adjusted some language to address this confusion.